File: backend/app/services/proposal_service.py
from datetime import datetime, timezone
from uuid import UUID

from sqlalchemy.exc import IntegrityError
from sqlalchemy.orm import Session

from app.core.exceptions import ConflictError, ForbiddenError, NotFoundError
from app.models.proposal import Proposal, ProposalStatus
from app.models.team import Team, TeamRole
from app.repositories.challenge_repository import ChallengeRepository
from app.repositories.membership_repository import MembershipRepository
from app.repositories.proposal_repository import ProposalRepository
from app.repositories.team_repository import TeamMembershipRepository, TeamRepository


class ProposalService:
    """Business logic for solution proposals (CP3 core + CP4 review).

    Owns transaction boundaries: repositories only flush; the service
    commits successful operations and rolls back on failure.

    Authorization is database-backed — team membership and institution
    membership (owner/representative) are resolved from the database at
    request time. Client-supplied role, status, ownership and membership
    data is never trusted.
    """

    def __init__(self, db: Session) -> None:
        self.db = db
        self.repository = ProposalRepository(db)
        self.team_repository = TeamRepository(db)
        self.team_membership_repository = TeamMembershipRepository(db)
        self.institution_membership_repository = MembershipRepository(db)
        self.challenge_repository = ChallengeRepository(db)

    # --- Authorization helpers (DB-backed) ----------------------------------

    def _get_active_membership(self, team_id: UUID, user_id: UUID):
        return self.team_membership_repository.get_active_membership(team_id, user_id)

    def _is_lead(self, team_id: UUID, user_id: UUID) -> bool:
        membership = self._get_active_membership(team_id, user_id)
        return membership is not None and membership.role == TeamRole.LEAD

    def _can_view(self, team: Team, user_id: UUID) -> bool:
        """Viewing is limited to active team members or active institution
        owners/representatives of the team's institution."""
        if self._get_active_membership(team.id, user_id) is not None:
            return True
        return self.institution_membership_repository.has_role(
            user_id, team.institution_id, ("owner", "representative")
        )

    def _can_review(self, team: Team, reviewer_user_id: UUID) -> bool:
        """A reviewer must be an ACTIVE owner or representative of the
        institution the proposal's team belongs to. Platform reviewers,
        students and ordinary team members never qualify."""
        return self.institution_membership_repository.has_role(
            reviewer_user_id, team.institution_id, ("owner", "representative")
        )

    def _require_reviewer(
        self, proposal: Proposal, reviewer_user_id: UUID
    ) -> None:
        """Ensure the user may review this proposal, raising 403 otherwise.

        The proposeal's institution is resolved from the database via its
        team — never from client input."""
        team = self.team_repository.get_by_id(proposal.team_id)
        if team is None:
            raise NotFoundError("Team", proposal.team_id)
        if not self._can_review(team, reviewer_user_id):
            raise ForbiddenError(
                "You do not have permission to review this proposal."
            )

    def _require_visible_proposal(self, proposal_id: UUID, user_id: UUID) -> Proposal:
        proposal = self.repository.get_by_id(proposal_id)
        if proposal is None:
            raise NotFoundError("Proposal", proposal_id)
        team = self.team_repository.get_by_id(proposal.team_id)
        if team is None or not self._can_view(team, user_id):
            raise ForbiddenError("You do not have access to this proposal.")
        return proposal

    # --- Create --------------------------------------------------------------

    def create_proposal(
        self,
        team_id: UUID,
        challenge_id: UUID,
        title: str,
        summary: str,
        approach: str | None,
        resources_needed: str | None,
        timeline: str | None,
        creator_user_id: UUID,
    ) -> Proposal:
        team = self.team_repository.get_by_id(team_id)
        if team is None:
            raise NotFoundError("Team", team_id)

        challenge = self.challenge_repository.get_by_id(challenge_id)
        if challenge is None:
            raise NotFoundError("Challenge", challenge_id)

        if team.challenge_id != challenge_id:
            raise ConflictError(
                "challenge_id does not match the challenge this team is working on."
            )

        if self._get_active_membership(team_id, creator_user_id) is None:
            raise ForbiddenError("You are not an active member of this team.")

        existing = self.repository.get_by_team_challenge(team_id, challenge_id)
        if existing is not None:
            raise ConflictError(
                "This team already has a proposal for this challenge."
            )

        try:
            proposal = self.repository.create(
                {
                    "team_id": team_id,
                    "challenge_id": challenge_id,
                    "title": title,
                    "summary": summary,
                    "approach": approach,
                    "resources_needed": resources_needed,
                    "timeline": timeline,
                    "status": ProposalStatus.DRAFT,
                }
            )
            self._commit()
        except ConflictError:
            raise
        except IntegrityError:
            self.db.rollback()
            raise ConflictError(
                "This team already has a proposal for this challenge."
            ) from None

        self.db.refresh(proposal)
        return proposal

    # --- Retrieval ------------------------------------------------------------

    def list_proposals(
        self,
        user_id: UUID,
        team_id: UUID | None = None,
        status: ProposalStatus | None = None,
        skip: int = 0,
        limit: int = 20,
    ) -> tuple[list[Proposal], int]:
        return self.repository.list_visible(
            user_id=user_id,
            team_id=team_id,
            status=status,
            skip=skip,
            limit=limit,
        )

    def get_proposal(self, proposal_id: UUID, user_id: UUID) -> Proposal:
        return self._require_visible_proposal(proposal_id, user_id)

    # --- Edit (draft only) --------------------------------------------------

    def update_proposal(
        self,
        proposal_id: UUID,
        user_id: UUID,
        fields: dict,
    ) -> Proposal:
        proposal = self.repository.get_by_id(proposal_id)
        if proposal is None:
            raise NotFoundError("Proposal", proposal_id)

        if self._get_active_membership(proposal.team_id, user_id) is None:
            raise ForbiddenError("You are not an active member of this team.")

        if proposal.status != ProposalStatus.DRAFT:
            raise ConflictError("Only draft proposals can be edited.")

        if not fields:
            return proposal

        try:
            updated = self.repository.update(proposal, fields)
            self._commit()
        except ConflictError:
            raise
        except IntegrityError:
            self.db.rollback()
            raise ConflictError("This proposal could not be updated.") from None

        self.db.refresh(updated)
        return updated

    # --- Submit (draft -> submitted, lead only) ------------------------------

    def submit_proposal(self, proposal_id: UUID, lead_user_id: UUID) -> Proposal:
        proposal = self.repository.get_by_id(proposal_id)
        if proposal is None:
            raise NotFoundError("Proposal", proposal_id)

        if not self._is_lead(proposal.team_id, lead_user_id):
            raise ForbiddenError("Only the active team lead can submit a proposal.")

        if proposal.status != ProposalStatus.DRAFT:
            raise ConflictError("Only draft proposals can be submitted.")

        proposal.status = ProposalStatus.SUBMITTED
        proposal.submitted_at = datetime.now(timezone.utc)
        self._commit()
        self.db.refresh(proposal)
        return proposal

    # --- Withdraw (draft|submitted -> withdrawn, lead only) ------------------

    def withdraw_proposal(self, proposal_id: UUID, lead_user_id: UUID) -> Proposal:
        proposal = self.repository.get_by_id(proposal_id)
        if proposal is None:
            raise NotFoundError("Proposal", proposal_id)

        if not self._is_lead(proposal.team_id, lead_user_id):
            raise ForbiddenError("Only the active team lead can withdraw a proposal.")

        if proposal.status == ProposalStatus.WITHDRAWN:
            raise ConflictError("This proposal has already been withdrawn.")

        if proposal.status not in (
            ProposalStatus.DRAFT,
            ProposalStatus.SUBMITTED,
        ):
            raise ConflictError("Only draft or submitted proposals can be withdrawn.")

        proposal.status = ProposalStatus.WITHDRAWN
        self._commit()
        self.db.refresh(proposal)
        return proposal

    # --- Review: submitted -> under_review -> accepted|rejected (CP4) --------
    # The review workflow is institution-owned: only an ACTIVE owner or
    # representative of the proposal's team institution may advance the state
    # machine. rejected/accepted are terminal — there is no resubmission,
    # appeal or second-round review. reviewed_at and reviewed_by are set
    # server-side from the authenticated reviewer at the final decision.

    def start_review(
        self, proposal_id: UUID, reviewer_user_id: UUID
    ) -> Proposal:
        proposal = self.repository.get_by_id(proposal_id)
        if proposal is None:
            raise NotFoundError("Proposal", proposal_id)

        self._require_reviewer(proposal, reviewer_user_id)

        if proposal.status != ProposalStatus.SUBMITTED:
            raise ConflictError(
                "Only submitted proposals can be moved into review."
            )

        proposal.status = ProposalStatus.UNDER_REVIEW
        self._commit()
        self.db.refresh(proposal)
        return proposal

    def accept_proposal(
        self,
        proposal_id: UUID,
        reviewer_user_id: UUID,
        review_note: str | None = None,
    ) -> Proposal:
        proposal = self.repository.get_by_id(proposal_id)
        if proposal is None:
            raise NotFoundError("Proposal", proposal_id)

        self._require_reviewer(proposal, reviewer_user_id)

        if proposal.status != ProposalStatus.UNDER_REVIEW:
            raise ConflictError(
                "Only proposals under review can be accepted."
            )

        proposal.status = ProposalStatus.ACCEPTED
        proposal.reviewed_at = datetime.now(timezone.utc)
        proposal.reviewed_by = reviewer_user_id
        proposal.review_note = review_note
        self._commit()
        self.db.refresh(proposal)
        return proposal

    def reject_proposal(
        self,
        proposal_id: UUID,
        reviewer_user_id: UUID,
        review_note: str | None = None,
    ) -> Proposal:
        proposal = self.repository.get_by_id(proposal_id)
        if proposal is None:
            raise NotFoundError("Proposal", proposal_id)

        self._require_reviewer(proposal, reviewer_user_id)

        if proposal.status != ProposalStatus.UNDER_REVIEW:
            raise ConflictError(
                "Only proposals under review can be rejected."
            )

        proposal.status = ProposalStatus.REJECTED
        proposal.reviewed_at = datetime.now(timezone.utc)
        proposal.reviewed_by = reviewer_user_id
        proposal.review_note = review_note
        self._commit()
        self.db.refresh(proposal)
        return proposal

    def _commit(self) -> None:
        try:
            self.db.commit()
        except Exception:
            self.db.rollback()
            raise

File: frontend/src/services/proposalService.js
/**
 * Proposal Service — communication with the FastAPI Proposal Engine (CP3).
 * Read surfaces only for now (workspace summaries, proposal detail).
 * Create / edit / submit / withdraw arrive in later slices.
 */

import { apiRequest } from "./api.js";

function buildQuery(params) {
  const search = new URLSearchParams();
  Object.entries(params).forEach(([key, value]) => {
    if (value === undefined || value === null || value === "") return;
    if (Array.isArray(value)) {
      if (value.length) search.set(key, value.join(","));
    } else {
      search.set(key, String(value));
    }
  });
  const qs = search.toString();
  return qs ? `?${qs}` : "";
}

/**
 * List proposals visible to the authenticated user (active team member,
 * or institution_admin/representative of the team's institution).
 * @param {Object} [params]
 * @param {string} [params.teamId]  Narrow to a single team
 * @param {string} [params.status]  ProposalStatus value ("draft"|"submitted"|...)
 * @param {number} [params.skip]
 * @param {number} [params.limit]
 */
export async function listProposals(params = {}) {
  const { teamId, status, skip = 0, limit = 20 } = params;
  return apiRequest(
    `/api/proposals${buildQuery({ team_id: teamId, status, skip, limit })}`
  );
}

/**
 * Fetch a single proposal's full detail (summary, approach, resources,
 * timeline, review fields).
 */
export async function getProposal(id) {
  if (!id) throw new Error("Proposal ID is required");
  return apiRequest(`/api/proposals/${encodeURIComponent(id)}`);
}

/**
 * Create a draft proposal for a team and challenge. Any ACTIVE team member
 * may create; the challenge must match the team's challenge (409 otherwise).
 * Proposals always start as a draft.
 * @param {{team_id: string, challenge_id: string, title: string, summary: string,
 *          approach?: string, resources_needed?: string, timeline?: string}} payload
 */
export async function createProposal(payload) {
  return apiRequest("/api/proposals", {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

/**
 * Edit a draft proposal (any ACTIVE team member). Only drafts are editable.
 * @param {string} id
 * @param {{title?: string, summary?: string, approach?: string|null,
 *          resources_needed?: string|null, timeline?: string|null}} fields
 *        Partial patch; title/summary must be non-blank when supplied.
 */
export async function updateProposal(id, fields) {
  return apiRequest(`/api/proposals/${encodeURIComponent(id)}`, {
    method: "PATCH",
    body: JSON.stringify(fields),
  });
}

/**
 * Submit a draft proposal (draft -> submitted). Team lead only;
 * submitted_at is set server-side.
 */
export async function submitProposal(id) {
  return apiRequest(`/api/proposals/${encodeURIComponent(id)}/submit`, {
    method: "POST",
  });
}

/**
 * Withdraw a draft or submitted proposal (-> withdrawn). Team lead only.
 * Withdrawal is terminal in CP3.
 */
export async function withdrawProposal(id) {
  return apiRequest(`/api/proposals/${encodeURIComponent(id)}/withdraw`, {
    method: "POST",
  });
}

/**
 * Advance the proposal review workflow (CP4). Institution admin or
 * representative of the proposal's team institution only.
 * @param {string} id
 * @param {{action: "start_review"|"accept"|"reject", review_note?: string}} payload
 *        - start_review: submitted -> under_review
 *        - accept:       under_review -> accepted
 *        - reject:       under_review -> rejected (review_note is captured)
 */
export async function reviewProposal(id, payload) {
  return apiRequest(`/api/proposals/${encodeURIComponent(id)}/review`, {
    method: "POST",
    body: JSON.stringify(payload),
  });
}